Fixes #11444 - Correctly count hosts in domain

tbrisker opened the pull request:
Only hosts that have a primary nic assigned to a domain should be
counted in the domain host count. Previous implementation was flawed in
that it only updated the counter in certain occasions and not all,
leading sometimes to incorrect host counts.
I have implemented the counter updates in `after_commit` to reduce
chance of deadlocks.

tbrisker committed:
Fixes #11444 - Correctly count hosts in domain

Only hosts that have a primary nic assigned to a domain should be
counted in the domain host count. Previous implementation was flawed in
that it only updated the counter in certain occasions and not all,
leading sometimes to incorrect host counts.
I have implemented the counter updates in `after_commit` to reduce
chance of deadlocks.

domcleal commented:
Please note that this PR needs rebasing.

domcleal commented on app/models/nic/base.rb:
```diff
@@ -223,6 +220,36 @@ def mac_uniqueness
+          end
+        end
+      else
+        return unless domain_changed && primary?
```
`elsif primary? && domain_changed` above?

domcleal commented on app/models/nic/base.rb:
```diff
@@ -223,6 +220,36 @@ def mac_uniqueness
+          domain.increment!('total_hosts') if domain_id.present?
+        else #changed from primary to non-primary
+          if domain_changed
+            Domain.find(previous_changes['domain_id'][0]).decrement!('total_hosts') if previous_changes['domain_id'][0].present?
```
Does this need to be `Domain.unscoped.find`, to bypass taxonomies in the default scope?

(ditto below)

domcleal commented:
Thanks, it looks okay but the amount of custom counting code is a little worrying.  Would you mind adding some more test coverage, particularly for the non-primary interfaces, primary to non-primary and vice versa cases?

tbrisker commented:
@domcleal I addressed your inline comments and added a fair bit of testing. I agree this is a lot of custom counting code. I'm starting to reconsider my stance on using [counter_culture](https://github.com/magnusvk/counter_culture) for the counters, both because of this and because of the various deadlock issues popping up around counters.

domcleal commented on app/models/domain.rb:
```diff
@@ -30,7 +30,7 @@ class Domain < ActiveRecord::Base
 
   scoped_search :on => [:name, :fullname], :complete_value => true
-  scoped_search :on => :hosts_count
+  scoped_search :on => :total_hosts
```
This should be aliased back to the original name.

domcleal commented on lib/tasks/fix_cached_counters.rake:
```diff
@@ -1,10 +1,15 @@
     puts "#{cl} corrected"
   end
+  Domain.all.each do |el|
+    Domain.reset_counters(el.id, :hostgroups)
```
hosts?

domcleal replied:
Oh nevermind, I get it now.  I didn't even know this had a hostgroups count.  Why!?

tbrisker replied:
Actually I have no idea, and I think this isn't the only counter that is useless. I think it can be removed, but I would suggest that it be done as a different PR.

domcleal commented:
Thanks, the test coverage is better now.  The only comment is the one about keeping the search keyword.

The gem you suggested does appear to fix a few of the problems that we've encountered.  I'm happy to get these tests and fixes in now though to enable refactoring later.

tbrisker commented:
@domcleal I added the alias. I think we should look into counter_culture again after the rails 4 upgrade if we still run into these problems, as the counter_cache handling in rails 4 is a bit better.

domcleal commented:
Merged as a152a1b, thanks @tbrisker.

tbrisker commented:
Thanks @domcleal, I have opened two issues as per our discussion:
[#12484](http://projects.theforeman.org/issues/12484) - Remove useless counters
[#12485](http://projects.theforeman.org/issues/12485) - Consider counter_culture gem
